boa_parser/src/lexer/mod.rs

```diff
@@ -272,17 +281,25 @@ impl<R> Lexer<R> {
                 '#' => PrivateIdentifier::new().lex(&mut self.cursor, start, interner),
                 '/' => self.lex_slash_token(start, interner),
                 '=' | '*' | '+' | '-' | '%' | '|' | '&' | '^' | '<' | '>' | '!' | '~' | '?' => {
-                    Operator::new(next_ch as u8).lex(&mut self.cursor, start, interner)
+                    Operator::new(
+                        next_ch
+                            .try_into()
+                            .expect("an ascii char must not fail to convert"),
+                    )
+                    .lex(&mut self.cursor, start, interner)
                 }
                 '\\' if self.cursor.peek()? == Some(b'u') => {
                     Identifier::new(c).lex(&mut self.cursor, start, interner)
                 }
                 _ if Identifier::is_identifier_start(c as u32) => {
                     Identifier::new(c).lex(&mut self.cursor, start, interner)
                 }
-                _ if c.is_ascii_digit() => {
-                    NumberLiteral::new(next_ch as u8).lex(&mut self.cursor, start, interner)
-                }
+                _ if c.is_ascii_digit() => NumberLiteral::new(
+                    next_ch
+                        .try_into()
+                        .expect("an ascii char must not fail to convert"),
+                )
+                .lex(&mut self.cursor, start, interner),
```

Comment:
I'm not sure about the `try_into`s here. A panic is for sure better from a safety perspective, but in these cases we can be sure that the conversion is safe. The panics also introduce a cold branch that is possibly bad for optimizations.

The safe solution without any panics would be to write out all the cases and use `b''` literals in each match branch. But I'm not sure how that would impact code generation as we then would have more different match arms.

Just some thoughts I had. I would also be fine merging this and maybe revisiting it later. What do you think?

Reply:
Fair enough :) 
I added some allows instead